### DISCUSSION TITLE
Mesa RL

### DISCUSSION THREAD

**Discussion by harshmahesheka** [mesa/mesa#2070] [Ideas]
Mesa provides a series of amazing multi-agent environments with various objectives. A natural use case of this is in Multi-Agent Reinforcement Learning. This year's  GSoC project is on a similar idea - [visit here](https://github.com/projectmesa/mesa/wiki/Projects-for-Google-Summer-of-Code-2024#mesa-rl). This thread is open to discuss how to go forward with this.

For getting started, I have implemented a few examples of MARL with Mesa using Stable Baselines and RLlib [here](https://github.com/harshmahesheka/mesa_rl). Other than the environments I think the visualization tools that mesa offers can really be helpful for RL tasks. Building on this learning, I propose wrappers around the existing Mesa Framework, allowing it to work directly with popular MARL libraries like RLlib. This approach avoids modifying existing Mesa implementations and makes it easier to develop and maintain. This concept aligns with my experience from the previous GSoC project, where I worked on similar functionalities. We could even include various trained model examples to give users a good starting point. Also, I've been exploring Petting Zoo, where they convert environments to RLlib-compatible formats using function calls. We can do a similar implementation for our case.

I'd love to hear everyone's feedback and thoughts on the project and idea!

**Comment by EwoutH** [comment]
Thanks for the write up. I agree with the wrapper approach. Most important for me would be a clean user-facing API to use it, together with solid documentation.

**Reply by harshmahesheka** [reply]
I read through Petting Zoo's documentation. Petting Zoo offers two APIs, Parallel and AEC, which provide different approaches for MARL tasks. Additionally, it provides a plethora of functions to interact with its extensive collection of environments, which are compatible with all major MARL libraries. Leveraging Petting Zoo's functions and open-source base can greatly benefit our project. There is a series of third part environments based on the petting zoo mentioned [here](https://pettingzoo.farama.org/environments/third_party_envs/), and we can do something similar.

Here's a possible plan:

Integration with Petting Zoo: We'll create a generalized wrapper to convert our environments into Petting Zoo environments. Then we use this wrapper and write code on top of it to create specific environment files with proper reward function action space and observation space for the environment as examples to be used directly.
 

Enhanced Visualization with Mesa Tools: We can improve our training and inference processes by integrating Mesa visualisation tools. These tools will offer insights and aid in analyzing agent behaviour during training and evaluation. We also leverage Petting Zoo's functions to enhance our environments with additional features. 

Training and Documentation: After converting and enhancing our environments, we'll train them using various algorithms supported by Petting Zoo. We'll also provide thorough documentation to ensure easy adoption by other users, making our trained policies accessible to a wider audience.

By following this approach, we'll be able to make the most of Petting Zoo's API and its open-source ecosystem while ensuring compatibility with various MARL libraries instead of just one. This will allow us to maximize the effectiveness of our environments. This is just a idea and your everyone's feedback on it will be great.

**Reply by EwoutH** [reply]
This AI-summary sounds nice on a high level. I would need more specifics to judge it further. I think minimal usage examples/demos are now the next steps.

**Reply by harshmahesheka** [reply]
Sure, I will work on training some more examples. This was to get started on a possible idea for a proposal. Do you have suggestions for any specific environment  I should work next with?

**Reply by EwoutH** [reply]
Nothing specific, I would go for something small that shows the functionality well.

**Comment by rht** [comment]
What are the alternatives to Petting Zoo, and why did you choose Petting Zoo over them? For context, we did an exhaustive compare/contrast for choosing the next gen framework used for Mesa frontend (Streamlit, Voila, panel, ipywidgets, Solara, gradio.app, ... https://github.com/projectmesa/mesa/discussions/1622), and for choosing the way to speed up Mesa (Numba, Cython, Rust, ... https://github.com/projectmesa/mesa/discussions/1610)

**Reply by harshmahesheka** [reply]
There are a few other options, such as Open Speil, but Petting Zoo is the most standard and is widely used by the community. It has an easy-to-integrate API and the largest set of compatible libraries. Moreover, I have not decided which one to use, so I will make a detailed comparison soon.
I discussed above whether it would make sense to use libraries like RLlib with Mesa directly or convert Mesa environments to MARL environments like PettingZoo. I have already done 2 examples of the first case; I am working on an example of 2nd method. Once we decide approach, I can do a detailed analysis of which framework to choose.

**Comment by harshmahesheka** [comment]
I have added an example implementation of converting a mesa environment into a petting zoo environment and using RLlib with it. What do you think about which of the two approaches makes more sense? Directly using Mesa with RL frameworks or converting it to environments like Petting Zoo.

**Reply by rht** [reply]
The [Petting Zoo environment](https://github.com/harshmahesheka/mesa_rl/blob/master/wolf_sheep_petting_zoo/environment.py) looks simpler and more modular to me, compared to the [RLlib one](https://github.com/harshmahesheka/mesa_rl/blob/master/wolf_sheep/model.py). But this is also because you have moved the helper functions in utils_function.py. What is the equivalent for the RLlib version's `get_observation` in the Petting Zoo version?

**Reply by harshmahesheka** [reply]
I think both implementations have the same get_observation function.

**Reply by EwoutH** [reply]
Thanks a lot for this, it's really exciting! I left some notes at https://github.com/harshmahesheka/mesa_rl/issues/1.

@quaquel I can highly recommend you to check this out.

**Comment by JJC022** [comment]
Did you evaluate the inclusion of other libraries like Coach or Gym? In particular, I'm curious what the interaction would look like with their existing visualization frameworks vis a vis Solara. I just finished my first RL class so my previous experience is with implementing Q-learning / Monte Carlo methods in Gym environments, so I'm just starting to dig into the RLib / Coach / PettingZoo documentation.

**Comment by EwoutH** [comment]
Competition news:
- https://github.com/JuliaDynamics/Agents.jl/pull/1170

We might be able to learn stuff from their approach.
